### PR TITLE
Feature/qi2 687/add OIDC

### PR DESCRIPTION
**Pull request by koffie**

### CODE REVIEW

File: quantuminspire/util/authentication.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+class OauthDeviceSession:
+    def __init__(self, settings: AuthSettings):
+        self._settings = settings
+        self._client_id = settings.client_id
+        self._token_info = settings.tokens
+        self._token_endpoint, self._device_endpoint = self._get_endpoints()
+        self._oauth_client = Client(settings.client_id)
+        self._headers = {"Content-Type": "application/x-www-form-urlencoded"}
+        self.expires_in = 600
+        self.interval = 5
```
Comment (NischalQuTech) [inline quantuminspire/util/authentication.py]: Please put comment about the units and also that the interval is polling interval

File: quantuminspire/util/authentication.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            "device_code": self._device_code,
+        }
+
+        response = requests.post(self._token_endpoint, data=data, headers=self._headers)
+        content = response.json()
+        if response.status_code == 400:
+            if content["error"] == "authorization_pending":
+                raise AuthorisationPending(content["error"])
+            if content["error"] == "slow_down":
+                self.interval += 5
```
Comment (NischalQuTech) [inline quantuminspire/util/authentication.py]: would be nice to make a variable for e.g. `self.delta_interval = 5 `

File: quantuminspire/util/authentication.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            except AuthorisationPending:
+                time.sleep(self.interval)
+
+        raise AuthorisationError("Login session timed out, please login again.")
+
+    def refresh(self) -> TokenInfo:
+        if self._token_info is None:
+            raise AuthorisationError("You should authenticate first before you can refresh")
+
+        if self._token_info.expires_at > time.time() + 5:
```
Comment (NischalQuTech) [inline quantuminspire/util/authentication.py]: also here, would be nice to have something like `self.token_window = 5`

File: quantuminspire/util/configuration.py [hunk 4]
```diff
@@ -34,22 +50,60 @@ def get_field_value(self, field: FieldInfo, field_name: str) -> Tuple[Any, str,
...
+    """A pydantic model for storing all information regarding oauth access and refresh tokens."""
+
+    access_token: str
+    expires_in: int
+    refresh_token: str
+    refresh_expires_in: int
+    generated_at: float = Field(default_factory=time.time)
+
+    @property
+    def expires_at(self) -> float:
```
Comment (NischalQuTech) [inline quantuminspire/util/configuration.py]: Can you make this explicit `access_expires_at`

File: quantuminspire/util/authentication.py (hunk 2 above)
Comment (koffie) [reply quantuminspire/util/authentication.py]: No this should not be variable. See: https://datatracker.ietf.org/doc/html/rfc8628
```
   slow_down
      A variant of "authorization_pending", the authorization request is
      still pending and polling should continue, but the interval MUST
      be increased by 5 seconds for this and all subsequent requests.
```

File: quantuminspire/util/authentication.py (hunk 1 above)
Comment (koffie) [reply quantuminspire/util/authentication.py]: done

File: quantuminspire/util/authentication.py (hunk 3 above)
Comment (koffie) [reply quantuminspire/util/authentication.py]: done

File: quantuminspire/util/configuration.py (hunk 4 above)
Comment (koffie) [reply quantuminspire/util/configuration.py]: done

**Review by NischalQuTech — APPROVED** [review]

**Pull request merged**